### DISCUSSION TITLE
dp_ipi outputs different energy from dp test

### DISCUSSION THREAD

**Discussion by JunbangLiu** [deepmodeling/deepmd-kit#1351] [Q&A] [answered]
I ran a training on a MoS2 monolayer and the trained model performs well when testing with dp test, but the MD simulation of i-pi with dp_ipi is not correct. I checked cell, atoms type and positions inputs and found nothing wrong. Here is some energy values outputed by dp_ipi(left) and dp test(right):
```
# deepmd_data_test: data_e pred_e
-2.241728591918945312e+01 -2.243821854646336789e+01   
-2.241053009033203125e+01 -2.243581020971576123e+01   
-2.240374374389648438e+01 -2.243328484084656083e+01
-2.239696502685546875e+01 -2.243066661421217844e+01
-2.239024162292480469e+01 -2.242795867697295620e+01
...
-2.250202560424804688e+01 -2.240865357210801179e+01
-2.251593399047851562e+01 -2.240791685812042999e+01
-2.252988243103027344e+01 -2.240669366386691053e+01
-2.254378128051757812e+01 -2.240502503234688092e+01
-2.255754089355468750e+01 -2.240287390496695252e+01
-2.257107162475585938e+01 -2.240011498705036530e+01
-2.258428001403808594e+01 -2.239672889577894210e+01
-2.259707641601562500e+01 -2.239274915570187829e+01
-2.260937309265136719e+01 -2.238808770543290194e+01
-2.262108039855957031e+01 -2.238279375773277380e+01
...
```
The differences seem not caused by unit conversion since they are quite big. And the structure is not destroyed during simulation so I think my inputs are basically right. I don't know how to deal with this problem, and any help will be appreciated. The trained model and ipi input files are in the zip files.
[dpipi.zip](https://github.com/deepmodeling/deepmd-kit/files/7697285/dpipi.zip)

**Comment by njzjz** [comment, marked as answer]
The cells in the `input.xml` and in the `box.npy` are not the same.

**Reply by JunbangLiu** [reply]
Actually they are the same. In i-pi manual it says 
> If ’mode’ is ’manual’, then ’cell’ takes a 9-elements vector containing the cell matrix (row-major). The 1st element define lattice vector a, the 2nd, 5th elements define lattice vector b, and the 3rd, 6th, 9th elements define lattice vector c. 

It seems there's a same mistake in dp_ipi codes? The cell matrix should be transposed. Nothing goes wrong if the cell is orthorhombic so you didn't find it. 
```
      // get box
      readbuffer_ (&socket, (char *)(cell_h),  9*sizeof(double));
      readbuffer_ (&socket, (char *)(cell_ih), 9*sizeof(double));
      for (int dd = 0; dd < 9; ++dd){
	dbox[dd] = cell_h[dd] * cvt_len;
      }
      region.reinitBox (&dbox[0]);
```

**Reply by njzjz** [reply]
Going to be fixed in #1353.

**Reply by JunbangLiu** [reply]
Unfortunately, I transposed `box.npy` and I still can't reproduce dp_ipi's energy values.
```
# deepmd_data_test: data_e pred_e
-2.241728591918945312e+01 -1.939457181877252623e+01
-2.241053009033203125e+01 -1.939495181045023742e+01
-2.240374374389648438e+01 -1.939532699656475501e+01
-2.239696502685546875e+01 -1.939563070342149942e+01
-2.239024162292480469e+01 -1.939538240339685871e+01
-2.238361740112304688e+01 -1.939503680863017721e+01
-2.237714385986328125e+01 -1.939466304302108313e+01
-2.237087059020996094e+01 -1.939419810260939414e+01
-2.236484718322753906e+01 -1.939306783428032332e+01
-2.235912704467773438e+01 -1.939188798575647255e+01
```
That is not, at least not only, the cause of my problem. Sorry.

**Reply by njzjz** [reply]
I am not sure, but after applying #1353, the energy is correct. The error is about 6e-6.

**Reply by JunbangLiu** [reply]
OK, waiting for the update, thanks.
